Core: Handle statistics file clean up from expireSnapshots

ajantha-bhat opened the pull request:
Currently, Statistics files are safeguarded against orphan_files cleanup. But they are never cleaned up from table metadata and from the storage once the snapshots are expired/deleted.

Hence, this PR adds a change to handle the Statistics file cleanup during expire_snapshot.

Note that this is just for API level clean up (`table#expireSnapshots`) 

Clean-up from expired snapshots spark action/procedure will be built on top of it in a follow-up PR.

ajantha-bhat commented:
cc: @findepi, @rdblue, @szehon-ho

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -79,4 +80,15 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
         .run(deleteFunc::accept);
   }
+
+  protected Set<String> expiredStatisticsFilesLocations(
+      TableMetadata beforeExpiration, Set<Long> expiredIds) {
+    Set<String> expiredStatisticsFilesLocations = Sets.newHashSet();
+    for (StatisticsFile statisticsFile : beforeExpiration.statisticsFiles()) {
+      if (expiredIds.contains(statisticsFile.snapshotId())) {
+        expiredStatisticsFilesLocations.add(statisticsFile.path());
+      }
+    }
+    return expiredStatisticsFilesLocations;
+  }
```
Curious, is it guaranteed that StatisticsFIles cannot be reused across multiple snapshots? 
The comment on StatisticsFile#snapshotId says 

`  /** ID of the Iceberg table's snapshot the statistics were computed from. */
  long snapshotId();
`
Is it possible that a StatisticsFile "file1" is computed at snapshot 1, but then still considered a valid reachable statistics file at Snapshot 2? and then we wouldn't want to do a removal of that file.

If it is possible it seems that we need to capture the reachability set rather than just removing all the statistics file before expiration. let me know if i misunderstood !

ajantha-bhat replied:
How do we support time travel if we reuse the same statistics file for all the snapshots?
Because of this, I assumed that each snapshot will create its own statistics file.

findepi replied:
@amogh-jahagirdar good point
sorry, didn't read this comment before posting https://github.com/apache/iceberg/pull/6090/files#r1013923155
@ajantha-bhat see there for an example where you want to reuse a stats file

ajantha-bhat replied:
@amogh-jahagirdar, @findepi  
As #6267 clarifies the spec. I have handled the reuse scenario and added the testcase now.

findepi commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -79,4 +80,15 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+    Set<String> expiredStatisticsFilesLocations = Sets.newHashSet();
+    for (StatisticsFile statisticsFile : beforeExpiration.statisticsFiles()) {
+      if (expiredIds.contains(statisticsFile.snapshotId())) {
+        expiredStatisticsFilesLocations.add(statisticsFile.path());
```
- `statisticsFile.snapshotId()` is "ID of the Iceberg table's snapshot the statistics were computed from",  it's not "ID of the snapshot the file is attached to"
- The spec allows one Puffin file to be attached to multiple snapshots. For example, rewrite_files / optimize will produce a new snapshot, but could carry forward some/all table-level stats.
  - So, the fact that the snapshot is being forgotten doesn't imply the stats file can be safely deleted.

ajantha-bhat replied:
> The spec allows one Puffin file to be attached to multiple snapshots. For example, rewrite_files / optimize will produce a new snapshot, but could carry forward some/all table-level stats.
So, the fact that the snapshot is being forgotten doesn't imply the stats file can be safely deleted.

makes sense. I will compute live statistics file set and then check whether it can be safely deleted along with the current changes.

ajantha-bhat replied:
@findepi: Thinking more about this, As the `TableMetadata` has just the list of  `StatisticsFile`. And you have mentioned,  statisticsFile.snapshotId() is "ID of the Iceberg table's snapshot the statistics were computed from"

So, how will the query knows which statistics file to use for the current snapshot (Incase of rewrite data files, the current snapshot id may not be present in that list of statistics file?)

ajantha-bhat commented:
> @findepi: Thinking more about this, As the TableMetadata has just the list of StatisticsFile. And you have mentioned, statisticsFile.snapshotId() is "ID of the Iceberg table's snapshot the statistics were computed from"
So, how will the query knows which statistics file to use for the current snapshot (Incase of rewrite data files, the current snapshot id may not be present in that list of statistics file?)

@rdblue, @findepi: Please help in clearing my above doubt.

findepi commented:
I think we should change the label of the `snapshot-id` entry in https://iceberg.apache.org/spec/#table-statistics (to level, not blob level)

ajantha-bhat commented:
> I think we should change the label of the snapshot-id entry in https://iceberg.apache.org/spec/#table-statistics (to level, not blob level)

Sorry, I still didn't get how the query engine will figure out the statistics file for the current snapshot (when the snapshot is reused).
Instead of the suggested change, can we change `statisticsFile.snapshotId()` to the snapshot id of the referring snapshot? This way `TableMetadata` will have entries for each snapshot id (even for the resue case). Snapshot file path can be reused. 

@rdblue: What do you think about this?

rdblue commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,33 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+            });
+      }
+    }
+    return expiredStatisticsFilesLocations;
```
Style: missing whitespace.

rdblue commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,33 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+      }
+    }
+
+    // Statistics files can be reused in case of rewrite data files.
```
Why would rewrite data files be involved?

ajantha-bhat replied:
I can remove/update the comment. 

But I just wanted to mention stats files can be reused (rewrite data files is one of the scenarios). So, we can't blindly remove it without checking the reachability graph.

rdblue commented on core/src/main/java/org/apache/iceberg/RemoveSnapshots.java:
```diff
@@ -204,7 +204,11 @@ private TableMetadata internalApply() {
-        .forEach(idsToRemove::add);
+        .forEach(
+            snapshotId -> {
+              updatedMetaBuilder.removeStatistics(snapshotId);
```
This should be handled inside of `removeSnapshots` because we `TableMetadata` that to maintain consistency.

ajantha-bhat replied:
Ah, true. I will move it.

rdblue commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,33 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
   }
+
+  protected Set<String> expiredStatisticsFilesLocations(
+      TableMetadata beforeExpiration, Set<Long> validIds, Set<Long> expiredIds) {
```
Rather than guessing which stats files were removed during expiration, I think that this should just diff the two lists of stats files.

ajantha-bhat replied:
ok. I will update it.

ajantha-bhat replied:
> just diff the two lists of stats files.

Two StatisticsFiles objects can use the same stats file location in case of reuse.  (for example during rewrite data files). I have simplified the logic now.

ajantha-bhat commented:
@rdblue, @findepi, @amogh-jahagirdar: Handled the comments. Please take a look at it again.    
Also, #6267 is ready.

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java:
```diff
@@ -264,6 +264,14 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
+    if (!beforeExpiration.statisticsFiles().isEmpty()) {
+      Set<String> expiredStatisticsFilesLocations =
+          expiredStatisticsFilesLocations(beforeExpiration, afterExpiration);
+      LOG.warn(
+          "Statistics files to delete: {}", Joiner.on(", ").join(expiredStatisticsFilesLocations));
```
I know it was warn level for the other cases in incremental file cleanup, but it seems like something we shouldn't keep following IMO because it's not really a warning case. We know we want to delete these files and it's intentional. 

I think this could be debug level or we can simply remove it.

ajantha-bhat replied:
I made it to debug and other places too I have changed to debug log.

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,25 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+  }
+
+  private Set<String> allStatisticsFilesLocation(TableMetadata tableMetadata) {
+    Set<String> allStatisticsFilesLocation;
+    if (tableMetadata.statisticsFiles() != null) {
+      allStatisticsFilesLocation =
+          tableMetadata.statisticsFiles().stream()
+              .map(StatisticsFile::path)
+              .collect(Collectors.toSet());
+    } else {
+      allStatisticsFilesLocation = Sets.newHashSet();
+    }
+    return allStatisticsFilesLocation;
```
Nit: I think we could just initialize `allStatisticsFile = Sets.newHashSet()`and remove the else block. Also newline after the block/before the return.

ajantha-bhat replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java:
```diff
@@ -260,10 +260,18 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
     deleteFiles(filesToDelete, "data");
-    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
-    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
+    LOG.debug("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
```
Joining should not be performed prior to passing the argument to debug, or else it will always create a string regardless of whether it is used.

rdblue commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,25 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+
+  protected Set<String> expiredStatisticsFilesLocations(
+      TableMetadata beforeExpiration, TableMetadata afterExpiration) {
+    Set<String> statisticsFilesLocationBeforeExpiry = allStatisticsFilesLocation(beforeExpiration);
```
The term ["expiry" should be "expiration"](https://books.google.com/ngrams/graph?content=epiry%2Cexpiration&year_start=1800&year_end=2019&corpus=26&smoothing=3).

rdblue commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,25 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+    Set<String> statisticsFilesLocationBeforeExpiry = allStatisticsFilesLocation(beforeExpiration);
+    Set<String> statisticsFilesLocationAfterExpiry = allStatisticsFilesLocation(afterExpiration);
+    statisticsFilesLocationBeforeExpiry.removeAll(statisticsFilesLocationAfterExpiry);
+    return statisticsFilesLocationBeforeExpiry;
```
I think it is a best practice to avoid mutating sets or returning mutable sets. Can you use Guava's `Sets.difference` instead of `removeAll`? It would also be better if the set returned by `allStatisticsFilesLocations` were immutable.

rdblue commented on core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java:
```diff
@@ -83,4 +85,25 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
+    return statisticsFilesLocationBeforeExpiry;
+  }
+
+  private Set<String> allStatisticsFilesLocation(TableMetadata tableMetadata) {
```
How about `statsFileLocations` instead? Since it returns a collection, the method name should be plural. And we also avoid needless modifiers, like `all` (which is assumed).

rdblue commented on core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java:
```diff
@@ -260,10 +260,18 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
+      Set<String> expiredStatisticsFilesLocations =
+          expiredStatisticsFilesLocations(beforeExpiration, afterExpiration);
+      LOG.debug(
+          "Statistics files to delete: {}", Joiner.on(", ").join(expiredStatisticsFilesLocations));
```
What is the value of this debug message?

ajantha-bhat replied:
Yes. It is only present in incremental flow (not reachable flow). I just followed the same pattern which is present for manifest, manifest list files. Shall I remove them too?

rdblue commented on core/src/main/java/org/apache/iceberg/TableMetadata.java:
```diff
@@ -1211,6 +1211,7 @@ public Builder removeSnapshots(Collection<Long> idsToRemove) {
         if (idsToRemove.contains(snapshotId)) {
           snapshotsById.remove(snapshotId);
           changes.add(new MetadataUpdate.RemoveSnapshot(snapshotId));
+          removeStatistics(snapshotId);
```
Can you explain why this PR is changing the behavior of `TableMetadata`?

ajantha-bhat replied:
expire snapshots used to remove the expired snapshot entires from `TableMetadata` (which is changing contents of `TableMetadata`).
So, now that those snapshots are expired. We need to clean up the statistics files entires present for those expired snapshots. Hence, modifying the  `TableMetadata`

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1515,4 +1594,63 @@ private RemoveSnapshots removeSnapshots(Table table) {
+                  .collect(ImmutableList.toImmutableList()));
+    }
+
+    commitStats(table, statisticsFile);
```
Functions should not have unexpected side-effects, like committing to a table when the expectation is to only write a file. Can you remove this?

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1515,4 +1594,63 @@ private RemoveSnapshots removeSnapshots(Table table) {
     return (RemoveSnapshots) removeSnapshots.withIncrementalCleanup(incrementalCleanup);
   }
+
+  private StatisticsFile writeStatsFileForCurrentSnapshot(Table table, File statsLocation)
```
I don't think there's a reason to pass table to this method. I think this should accept a String location, a FileIO, and a snapshot ID.

This should also not use `File` for writing.

ajantha-bhat replied:
This was taken from existing code.
https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java#L917

I will modify the current testcase.

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1245,74 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+
+    Assert.assertEquals("Should have 2 statistics file", 2, table.statisticsFiles().size());
+
+    table.updateProperties().set(TableProperties.MAX_SNAPSHOT_AGE_MS, "1").commit();
```
Expiring based on assumptions about time possibly introduces flakiness to this test.

The right way to test this is to use `long afterFirstSnapshot = waitUntilAfter(table.currentSnapshot().timestampMillis()` and `table.expireSnapshots().expireOlderThan(afterFirstSnapshot).commit()`

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1245,74 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+
+    table.newAppend().appendFile(FILE_B).commit();
+    // reuse the existing stats file with the current snapshot
+    StatisticsFile statisticsFile2 = reuseStatsForCurrentSnapshot(table, statisticsFile1);
```
I can understand why you'd want to test that reused files are not removed. However, this metadata is invalid so I think this test should have a warning comment at the top about why it is testing this case and that it is not expected in real tables.

ajantha-bhat commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1515,4 +1594,63 @@ private RemoveSnapshots removeSnapshots(Table table) {
+    transaction.commitTransaction();
+  }
+
+  private File statsFileLocation(Table table) throws URISyntaxException {
```
This was taken from existing code.
https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java#L910-L915

I will modify the current testcase.

ajantha-bhat commented:
@rdblue :  Thanks for the awesome review. I will keep those points in mind. 

I have addressed all the comments. Please take a look at it again.

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1243,95 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+        statisticsFile1.snapshotId());
+
+    table.newAppend().appendFile(FILE_B).commit();
+    // Note: In the real scenario, appendFile will always create a new statistics file.
+    // Only rewrite_data_files scenario can reuse the statistics file with new snapshot id.
+    // To avoid simulating rewrite, this test is reusing the stats for append operation itself.
```
Just a nit, I think this inline comment could be a bit more concise, maybe something like:

"RewriteDataFiles can reuse statistics files across operations. This test reuses stats for append just to mimic this scenario without having to run RewriteDataFiles."

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java:
```diff
@@ -260,10 +259,14 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
     deleteFiles(manifestsToDelete, "manifest");
     deleteFiles(manifestListsToDelete, "manifest list");
+
+    if (!beforeExpiration.statisticsFiles().isEmpty()) {
+      Set<String> expiredStatisticsFilesLocations =
+          expiredStatisticsFilesLocations(beforeExpiration, afterExpiration);
+      deleteFiles(expiredStatisticsFilesLocations, "statistics files");
+    }
```
Nit: I think we could just have expireStatisticsFilesLocation short circuit before computing  in case before expiration is empty. Rather than have the if(!beforeExpiration.statisticsFiles().isEmpty()) in both cleanup implementations

amogh-jahagirdar reviewed (approved):
LGTM some minor comments, but nothing blocking from my perspective.

ajantha-bhat commented:
ping @rdblue

ajantha-bhat commented:
Just another reminder.

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1243,95 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+            table.currentSnapshot().sequenceNumber(),
+            statsFileLocation1,
+            table.io());
+    commitStats(table.newTransaction(), statisticsFile1);
```
There's no need to use transactions for a single operation. Can you remove this and update the table directly?

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1515,4 +1613,51 @@ private RemoveSnapshots removeSnapshots(Table table) {
     return (RemoveSnapshots) removeSnapshots.withIncrementalCleanup(incrementalCleanup);
   }
+
+  private StatisticsFile writeStatsFileForCurrentSnapshot(
```
This name doesn't make sense. The snapshot ID and sequence number are passed in. I think it should be `writeStatsFile`.

ajantha-bhat replied:
ACK. Missed renaming during refactoring.

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1243,95 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+    Assert.assertEquals("Should have 2 statistics file", 2, table.statisticsFiles().size());
+
+    long tAfterCommits = waitUntilAfter(table.currentSnapshot().timestampMillis());
+    removeSnapshots(table).expireOlderThan(tAfterCommits).commit();
```
Please add a comment about why the second snapshot and file aren't removed (they are the current table state).

ajantha-bhat replied:
added a comment below

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1243,95 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+    table.newAppend().appendFile(FILE_B).commit();
+    // Note: RewriteDataFiles can reuse statistics files across operations.
+    // This test reuses stats for append just to mimic this scenario without having to run
+    // RewriteDataFiles.
```
Does this actually happen in `RewriteDataFiles`? I don't think that the same stats file should be added more than once. It's a good idea to make sure it doesn't, but that should not be the behavior of built-in operations.

ajantha-bhat replied:
@findepi has mentioned about reusing the stats file. I think we should not allow it because concurrent operations can add extra stats during rewrite operation.   

We don't have any engine integration with stats in this repo. So, I mentioned "can reuse" instead of "will resue"

rdblue replied:
I think this is inaccurate then. It should be enough to state that in the even that a snapshot file is for some reason reused, we want to detect that it is still referenced and not delete it from the file system.

ajantha-bhat replied:
ok. rephrased it.

rdblue commented:
Thanks, @ajantha-bhat! I made some comments in tests to fix.

ajantha-bhat commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1234,6 +1243,95 @@ public void testMultipleRefsAndCleanExpiredFilesFailsForIncrementalCleanup() {
+            statsFileLocation1,
+            table.io());
+    commitStats(table.newTransaction(), statisticsFile1);
+    Assert.assertEquals(
```
removed in all the places

ajantha-bhat commented:
> Thanks, @ajantha-bhat! I made some comments in tests to fix.

Thanks for the review. I have addressed the comments.

ajantha-bhat commented:
If the changes are ok, please merge this PR. So that I can rebase #6091 and make it ready for review.

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1515,4 +1599,47 @@ private RemoveSnapshots removeSnapshots(Table table) {
+    }
+  }
+
+  private StatisticsFile reuseStatsForCurrentSnapshot(
```
This is not for the "current" snapshot because the snapshot ID is being passed in.

When there are problems that need to be fixed in multiple places, I might just mention it once to avoid unnecessary repetition. So to keep PRs moving faster, you should always look for similar cases that also need to be fixed.

ajantha-bhat replied:
ACK. 

Apologies for the back and forth. This was induced during refactoring.

rdblue reviewed (changes requested):
@ajantha-bhat, looks like there are just two more things to fix. Thanks!

ajantha-bhat commented:
> @ajantha-bhat, looks like there are just two more things to fix. Thanks!

Done. Thanks for the review.

ajantha-bhat commented:
@jackye1995: Can you please consider this for the 1.2.0 release?

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1515,4 +1599,46 @@ private RemoveSnapshots removeSnapshots(Table table) {
+  }
+
+  private void commitStats(Table table, StatisticsFile statisticsFile) {
+    table.updateStatistics().setStatistics(statisticsFile.snapshotId(), statisticsFile).commit();
```
I find it odd that a single line like this is in a separate method. Seems like this could be inlined and would make the tests more readable.

rdblue commented:
Thanks, @ajantha-bhat. Good to have this in.

ajantha-bhat commented:
@rdblue: Thanks for the review and merge.   

Now, I have rebased and reworked #6091  based on the learnings of this PR. 
So, it is ready for review.
